pydantic/main.py

```diff
@@ -937,54 +935,11 @@ def __init__(self, __type: Any, *, config: CoreConfig | None = None) -> None:
 
     def __init__(self, __type: Any, *, config: CoreConfig | None = None) -> None:
         self._type = __type
-        merged_config: CoreConfig = {
-            **(config or {}),  # type: ignore[misc]
-            **getattr(__type, 'model_config', {}),
-        }
-        arbitrary_types = bool((config or {}).get('arbitrary_types_allowed', False))
-        local_ns = _typing_extra.parent_frame_namespace(parent_depth=2)
-        # BaseModel uses it's own __module__ to find out where it was defined
-        # and then look for symbols to resolve forward references in those globals
-        # On the other hand Validator() can be called with arbitrary objects,
-        # including type aliases where __module__ (always `typing.py`) is not useful
-        # So instead we look at the globals in our parent stack frame
-        # This works for the case where Validator() is called in a module that
-        # has the target of forward references in its scope but
-        # does not work for more complex cases
-        # for example, take the following:
-        #
-        # a.py
-        # ```python
-        # from typing import List, Dict
-        # IntList = List[int]
-        # OuterDict = Dict[str, 'IntList']
-        # ```
-        #
-        # b.py
-        # ```python
-        # from pydantic import Validator
-        # from a import OuterDict
-        # IntList = int  # replaces the symbol the forward reference is looking for
-        # v = Validator(OuterDict)
-        # v({"x": 1})  # should fail but doesn't
-        # ```
-        #
-        # If OuterDict were a BaseModel this would work because it would resolve
-        # the forward reference within the `a.py` namespace.
-        # But Validator(OuterDict) can't know what module OuterDict came from.
-        # In other words, the assumption that _all_ forward references exist in the
-        # module we are being called from is not technically always true
-        # Although most of the time it is and it works fine for recursive models and such/
-        # BaseModel's behavior isn't perfect either and _can_ break in similar ways,
-        # so there is no right or wrong between the two.
-        # But at the very least this behavior is _subtly_ different from BaseModel's.
-        global_ns = sys._getframe(1).f_globals.copy()
-        global_ns.update(local_ns or {})
-        gen = _generate_schema.GenerateSchema(
-            arbitrary_types=arbitrary_types, types_namespace=global_ns, typevars_map={}
-        )
-        schema = gen.generate_schema(__type)
-        self._validator = SchemaValidator(schema, config=merged_config)
+        # TODO: clean this up somehow? should we just import this at the module level?
+        from pydantic.funcs import analyze_type
+
+        schema = analyze_type(__type, config=config, parent_depth=3)  # type: ignore
+        self._validator = SchemaValidator(schema.core_schema, config=schema.core_config)
```

Comment:
Not sure how we should deal with this and also what the relationship between `Validator` and the free-standing functions will be. Given that we have `analyze_types` I think there's something to be said for not introducing this thin abstraction over `SchemaValidator` and having users use `pydantic_core` directly? There's also now issues related to stack depth since I'm introducing another call here, it's making one of our tests fail...